오픈마켓 II [STEP 2] 아리, 제리

leeari95 opened the pull request:
안녕하세요. 그린 @GREENOVER 
아리 @leeari95, 제리 @llghdud921 입니다!

오픈마켓2 마지막 스텝 구현을 마쳐 PR을 보냅니다!
벌써 마지막 스텝이네요... 시간이.. 순삭됐어요...
이번에도 저희가 할 수 있는 만큼! 해보고싶은 부분 추가로 구현해서 보내드립니다~

혹시 부족한 점이나 놓친 부분이 있다면 번거롭더라도 한번 더 짚어주시면 감사하겠습니다.  🙇🏻‍♀️

마지막까지도 잘부탁드릴게요~~~ 😃

## 고민했던 점

### 1. 에러처리

- 사용자가 볼 필요가 없는 에러의 경우(네트워크 에러) 얼럿을 띄우는 대신 print를 호출하도록 하여, 에러처리를 해주었습니다.

### 2. Custom Font에도 Dynamic Type을 적용

- UIFont가 제공하는 preferredFont를 사용하면 따로 굵기를 지정할 수 없고, 지정된 font만 사용해야한다. 반대로 systemFont를 사용한다면 Dynamic Type이 동작하지 않는다.
    - extension을 통해서 굵기를 지정해도 Dynamic Type을 지원하는 메소드를 구현하여 이 문제를 해결해주었습니다.

### 3. 이미지 상세보기 시 보고있던 이미지가 그대로 넘어가도록 구현

- 다른 마켓앱을 살펴보면 상세보기로 넘어갈 때 보고있던 이미지를 그대로 넘겨서 상세보기 화면으로 전환되는 것을 확인할 수 있습니다. 사용자 입장에서도 보고있던 이미지를 그대로 상세보기로 보는 것이 바람직하다고 판단되어 해당 기능을 구현하였습니다.
    
    ```swift
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
        let indexPath = IndexPath(item: self.currentPage, section: 0)
        self.collectionView.scrollToItem(
            at: indexPath,
            at: [.centeredHorizontally, .centeredVertically],
            animated: false
         )
    }
    ```
    

### 4. HIG를 참고하여 Action Sheet를 설계

> **Make destructive choices prominent.** Use red for buttons that perform destructive or dangerous actions, and display these buttons at the top of an action sheet.
> 
- 이 부분을 참고하여, 요구사항처럼 삭제 버튼을 아래에 위치해있는게 아니라 상단에 위치하도록 개선하였습니다.

### 5. CollectionView, PageControl을 이용한 이미지 Paging

- 상세페이지화면에서 이미지를 표시할 때 CollectionView의 isPaging과 가로 스크롤을 통해서 `paging`을 구현하였습니다.
- pageControl을 이용하여 현재 페이지의 위치를 나타내고 PageControl의 dot을 탭하여도 페이지가 전환되는 기능을 구현하였습니다.

### 6. 이미지 나타낼때 DispatchGroup을 사용

- 메인페이지에서 상세페이지로 넘어갈때  이미지를 네트워킹하는 속도보다 화면이 전환되는 시점이 빨라서 이미지가 보이지 않는 현상이 나타났습니다.
- DispatchGroup을 이용하여 네트워킹하는 코드 내 completion 탈출 클로저가 모두 끝난 시점에 View를 띄워주도록 하였습니다.
    
    ```swift
    dispatchGroup.enter()
    ImageManager.shared.downloadImage(with: newImage.url) { image in
        ImageManager.shared.setCacheData(of: image, for: newImage.url)
        self.images.append(image)
        dispatchGroup.leave()
    }
    
    dispatchGroup.notify(queue: .main) {
        DispatchQueue.main.async {
            // view 
            ...
        }
    }
    ```
    

## 궁금했던 것 / 조언이 필요한 부분

- `DetailViewController` → `ImageDetailViewController` 로 전환할 때 viewDidLoad에서 collectionView.`scrollToItem`을 호출 시 제대로 적용되지 않아서 DispatchQueue.main.`asyncAfter`를 활용했는데, 적절한지 잘 모르겠습니다... 조언이 필요합니다!
- 네트워킹 하는 부분이 마치 산맥..처럼 공포의 들여쓰기가 생겨났는데, 이러한 부분을 어떻게 더 개선할 수 있을지 더이상 떠오르지가 않네요.. 저희가 할 수 있는 만큼 최대한 메소드를 분리해보았는데, 이대로도 괜찮을까요?
- CollectionView에 FlowLayout을 따로 코드로 대입해주지 않으면, UICollectionViewDelegateFlowLayout를 채택하여 레이아웃을 설정해주어도 레이아웃이 적용되지 않는데, 왜그런걸까요?? 조언 부탁드립니다!
- ScrollView를 활용해서 Zoom을 구현하였지만, 디바이스 전체 높이에 맞춰서 maximumZoomScale을 설정하는 방법을 모르겠습니다.
    - 이미지가 디바이스 크기보다 커지지 않도록 하고싶은데... 방법이 없을까요? 임의로 2.0으로 지정해주긴 하였는데.. 좀더 공부해보고 싶어요!
- 에러처리하는 부분을 일단 사용자에게 알릴 수 있는 얼럿을 제외하고는 네트워크 부분이나, 리퀘스트 부분은 print를 호출하도록 하였는데, 이 부분이 적절한걸까요? 부족하다면 조언 부탁드리겠습니다!
- requestDelete 메소드 내부에서 success시 해당 데이터를 가지고 또 이어서 delete 메소드를 호출하여 삭제 요청을 진행하는 부분이 적절한 흐름인지 잘모르겠습니다.
    - 일단 저희가 생각해보았을 때 같은 id를 삭제 Secret 조회후 삭제 요청을 하는 과정에서 에러가 날 것 같지는 않는다고 판단되어지는데, 만약 이부분에서 에러가 난다면 어떤 에러인걸까요...? 테스트를 할수도 없어서 이 부분을 이렇게 그냥 냅둬도 되는건지 잘 판단이 안서네요..
    - 
    
    ```swift
    private func requestDelete(secret: String, complection: @escaping (Bool) -> Void) {
        guard let request = requestSecretSearch(secret: secret) else {
            return
        }
        let network = Network()
        network.execute(request: request) { result in
            switch result {
            case .success(let value):
                guard let value = value,
                      let password = String(data: value, encoding: .utf8) else {
                    return
                }
                self.delete(password: password) { isSuccess in
                    if isSuccess {
                        complection(true)
                    } else {
                        complection(false) // 여기서.. false가 나올까요...??? 서버가 이상하다면...?
                    }
                }
            case .failure(let error):
                print(error.localizedDescription)
                complection(false)
            }
        }
    
    private func delete(password: String, complection: @escaping (Bool) -> Void) {
            guard let data = self.data else {
                return
            }
            let networkManager = NetworkManager()
            guard let requestDelete = networkManager.requestDelete(id: UInt(data.id), secret: password) else {
                return
            }
            networkManager.fetch(request: requestDelete, decodingType: Product.self) { result in
                switch result {
                case .success:
                    complection(true)
                case .failure(let error):
                    print(error.localizedDescription) // 이 부분에서 print를 호출해주긴 합니다.
                    complection(false)
                }
            }
        }
    }
    ```

leeari95 committed:
feat: DetailViewController 내부에 images를 표시해주는 CollectionView를 추가 #21

- ImageDetailCell 생성하여 이미지뷰 추가
- 기존에 IBOutlet으로 있던 imageView를 제거
- 이미지의 순서를 표시하기 위한 UIPageControl 추가
- setUpImages를 네트워크가 완료되면 호출되도록 수정
- setUpImages 내부에 DispatchGroup을 두어 Task를 추적하고 끝나면 CollectionView를 구성할 수 있도록 구현
- viewDidLoad시 NavigationItem, PageControl, CollectionView를 setUp하도록 구성
- CollectionView 추가에 따른 DataSource와 Delegate 채택
- Detail.storyboard를 수정하고 ImageDetailCell.xib를 추가
- setUpTitle을 추가하여 화면전환 시 상품의 이름으로 Detail View의 Title을 설정하도록 구성

llghdud921 committed:
feat: pageControl 클릭시에 페이지 넘어가는 기능 추가 #22

leeari95 committed:
feat: DetailViewController 내부에 게시글을 삭제하는 기능 추가 #22

- tappedEditButton에 deleteHandler 추가
- API에 삭제 Secret 조회, Delete 요청까지 시도하는 requestDelete 메소드 구현
- Secret 조회를 요청하는 requestSecretSearch 메소드 구현
- API에 Delete를 요청하는 delete 메소드 구현
- requestSecretSearch에 불필요한 파라미터 제거
- tappedEditButton 내부에 불필요한 secret 변수 부분을 제거하여 개선

llghdud921 committed:
feat: detailViewController내 productInfo를 나타내는 View 속성 구현 #21

- storyboard로 stackView 구성 및 폰트 설정
- ProductInfoStackView customView 구현
- ProductInfoStackView내 text 나타내는 기능 구현
- UIFont extension을 이용하여 bold체 적용

leeari95 committed:
refactor: MainViewController 네트워킹 부분 버그 개선 #23

- 더이상 데이터가 없어서 Cell을 불러오는 부분에서 index out of range 에러가 나지 않도록 개선
- resetProductList로 데이터를 모두 지우는 부분을 지우고 덮어쓰는 방식으로 수정
- Detail 내부에 StockLabel의 Priority를 수정

llghdud921 committed:
feat: 상세페이지 내 이미지 클릭시 상세이미지 표시 및 줌 기능 구현 #22

- DetailViewController내 상세이미지 나타내는 performsegue 및 관련 코드 구현
- ImageDetailViewController 생성, storyboard내 ImageDetailViewController 구현
- ImageDetailViewController내 scrollview의 zoomscale을 가지고 계산하여 scrollview origin에 적용하는 zoom기능 구현
- ImageDetailViewController내 tapGestureRecognizer를 이용한 Zoom handling 기능 구현

leeari95 committed:
refactor: tappedEditButton 메소드 내부를 리팩토링 #24

- 하드코딩 되어있는 문자열을 AlertMessage로 리팩토링
- 핸들러에 있는 각 작업들을 메소드로 분리하여 개선

llghdud921 committed:
refactor: DetailViewController내 CollectionView, PageControl 커스텀뷰로 정의 #24

leeari95 committed:
feat: 이미지 상세보기 시 다른 이미지도 쓸어서 상세히 볼 수 있도록 구현 #21 #24

- 기존 ImageDetailCell의 네이밍을 ImageViewCell로 변경
- ImageDetailViewController에 있던 ScrollView 대신에 CollectionView로 교체
- PageControl 추가하여 몇번째 이미지인지 표시되도록 구현
- ImageDetailViewController로 화면전환 됬을 때 보고있던 이미지로 이동하도록 기능 구현
- 2번 탭했을 때 줌 기능을 ImageDetailCell에 구현

llghdud921 committed:
refactor: 상세페이지 accessibilty 적용 및 접근제어자 수정 #24 #22

leeari95 committed:
style: 사용하지 않는 불필요한 메소드 제거

llghdud921 committed:
refactor: 사용자가 볼 필요가 없는 에러의 경우 print로 에러처리를 하도록 개선 #24

leeari95 committed:
docs: README.md 파일 STEP4 내용 추가

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/ImageDetailViewController.swift:
```diff
@@ -0,0 +1,119 @@
+                at: [.centeredHorizontally, .centeredVertically],
+                animated: false
+            )
+        }
```
asyncAfter로 딜레이 걸어주는건 위험합니다ㅠ
구동은 되지만 본질적인 해결은 아닙니다..!
view lifecycle을 먼저 잘 생각해보면 좋을것 같아요🤔
이 28~33 라인을 viewWillAppear에서 호출하는게 맞을것 같습니다🙋🏻

llghdud921 replied:
딜레이를 하는게 문제였군요!! 감사합니다 :)
viewWillAppear로 개선하였습니다~~

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImageDetailCell.swift:
```diff
@@ -0,0 +1,84 @@
+        scrollView.minimumZoomScale = 1.0
+        scrollView.showsVerticalScrollIndicator = false
+        scrollView.showsHorizontalScrollIndicator = false
+        scrollView.decelerationRate = .fast
```
오호.. 스크롤 감속까지 신경쓰셨네요👍

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImageDetailCell.swift:
```diff
@@ -0,0 +1,84 @@
+    private func setUpScrollView() {
+        scrollView.delegate = self
+        scrollView.zoomScale = 1.0
+        scrollView.maximumZoomScale = 2.0
+        scrollView.minimumZoomScale = 1.0
```
이 조건이 특정 요구사항이 없다면 흔히들 많이 쓰는 기본값이긴하죠🤔

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImageDetailCell.swift:
```diff
@@ -0,0 +1,84 @@
+            return
+        }
+        
+        let toScale = maxScale
```
이 상수는 필요 없을것 같습니다..!🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImageDetailCell.swift:
```diff
@@ -0,0 +1,84 @@
+        }
+        
+        let toScale = maxScale
+        let finalScale = currectScale == minScale ? toScale : minScale
```
여기서 단순히 현재 스케일이 최소 지정된 스케일이면 maxScale로 설정하고 그게 아니면 전부 minScale로 설정함으로
toScale 대신 maxScale로 충분히 표현될 수 있을것 같습니다🤔

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImageDetailCell.swift:
```diff
@@ -0,0 +1,84 @@
+        
+        if minScale == maxScale, minScale > 1 {
+            return
+        }
```
해당 메서드가 private하게 쓰이고 있는데 사실 이 조건을 탈 수 있는 경우는 없다고 생각해요🤔
이미 이전에 setUpScrollView에서 모두 세팅이 되었기에 그 세팅된 조건에서는 이 로직을 타는 일이 있을까요..?🧐
만약 해당 메서드를 Public하게 가져간다면 쓰임이 있을 수 있겠지만 지금으로썬 없어도 될 코드라고 생각이 들었습니다🙋🏻

leeari95 replied:
아.. 생각해보니 그렇네요. 사실 저희가 강의를 참고하여 만들다보니까.. 해당 코드가 불필요하게 추가된 것 같아요.
위 코멘트들도 확인하여 불필요한 부분은 지워주도록 하겠습니다.
피드백 감사합니다 😊

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -89,41 +105,126 @@ class DetailViewController: UIViewController {
+        }
+    }
+    
+    private func modifyAction(_ secret: String) {
```
어떤걸 수정하는지 `modifyProduct` 같은 좀 더 명확한 네이밍이면 좋을것 같습니다🙋🏻

llghdud921 replied:
modifyAction과 deleteAction은 내부 작동을 명확하게 알리도록 showModifyPage, ShowDeleteSuccess로 네이밍 개선하였습니다!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -89,41 +105,126 @@ class DetailViewController: UIViewController {
+        self.requestDelete(secret: secret) { isSuccess in
+            if isSuccess {
+                DispatchQueue.main.async {
+                    self.showAlert(message: AlertMessage.CompleteProductdelete) {
```
`CompleteProductDelete`로 카멜 케이스를 따라 나타내면 더 좋겠습니다🙋🏻

llghdud921 replied:
넵 수정했습니다!!!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ImagePageControl.swift:
```diff
@@ -0,0 +1,27 @@
+
+import UIKit
+
+class ImagePageControl: UIPageControl {
+
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        setUpPageControl()
+    }
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+
+    private func setUpPageControl() {
+        currentPage = 0
+        pageIndicatorTintColor = .lightGray
+        currentPageIndicatorTintColor = .black
+        isHidden = true
+    }
+}
```
크.. 애플스럽게 페이지컨트롤을 사용하셨네요!
고급지고 아주 익숙하고도 예쁜 구현 구현입니다👍

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ProductInfoStackView.swift:
```diff
@@ -0,0 +1,81 @@
+        setUpAccessibility()
+    }
+    
+    func setUpAccessibility() {
```
요 메서드는 private해도 될것 같습니다🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/View/ProductInfoStackView.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+    
+    private func setUpPrice(product: Product) {
+        let formatedPrice = "\(product.currency.rawValue) \(product.price.fomattedString)"
```
아마 포맷팅된 가격을 의미하는 `formattedPrice`를 의도하셨을것 같아요..!
저도 매번 네이밍때 헷갈리는게 format 과거형이 어렵더라구요ㅠ 
formated는 아예 다른 뜻으로 사용되고 있습니다.
(아마 코딩하면서 사용할 일은 없는 단어일것 같아요🥲)

llghdud921 replied:
잘못 적었네요...ㅎㅎㅎ 이전에 main에 구현된 코드도 잘못적어서 두개 다 수정완료했습니다~!

GREENOVER commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # 오픈 마켓
 
-1. 프로젝트 기간: 2022.01.03 - 2022.01.14
+1. 프로젝트 기간: 2022.01.03 - 2022.01.28
```
리드미 끝까지 완수해내셨네요..!👍👍👍
너무너무 고생많으셨습니다!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -89,41 +105,126 @@ class DetailViewController: UIViewController {
+        }
+    }
+    
+    private func requestDelete(secret: String, complection: @escaping (Bool) -> Void) {
```
대부분의 로직 코드에서 completion을 complection으로 오표기 되어있습니다🙋🏻

leeari95 replied:
앗.. 전부 찾아내서 수정하였습니다... 감사합니다 🙇🏻‍♀️

GREENOVER reviewed (changes requested):
@leeari95 @llghdud921 
🥳🥳🥳🥳🥳
여러분 스텝을 모두 완주 하신것 너무 축하드립니다!!
정말로 이렇게 빠르고 요구사항도 다 구현하면서 추가 기능까지 구현해주신게 아주 대단해요!
저도 그렇게 못했었는데..! 최고네요 정말👍
전체적으로 단순히 토이 프로젝트를 넘어 완성도 있는 앱이 하나 만들어 졌다고 생각해요!
그정도로 기능적 완성 측면에서는 만점입니다ㅎㅎ
그래서 이번 리뷰는 크게 드릴 말씀은 없었고 질문해주신 부분에 대해 의견이나 사소한 부분들을 남겨봤어요ㅎㅎ
다시 한번 고생하셨어요!
🥳🥳🥳🥳🥳
***
### 칭찬 드리고 싶은 부분
- 요구사항과 추가적으로 욕심내서 구현하신 부분들이 너무 디테일하고 완성도를 높였어요 💯
- 폴더와 파일을 구조적이고 명확하게 잘 나눠주셨어요 💯
- 리드미 드디어 결실을 맺으셨네요 고생하셨습니다! 💯
- UI의 여러 메서드 및 프로퍼티들을 잘 활용하시는점 최고네요 💯 (저도 처음 접하는것도 있어서 배우고 갑니다!🙇🏻)
***
### 고민하셨던 부분에 대한 저의 의견
1. 에러처리
- 아하 그렇게 변경하셨군요! 네 그 방법도 좋은것 같습니다ㅎㅎ
실패 후 후속으로 재통신을 해준다던가 하는 작업만 있다면 충분히 좋은 코드같습니다.
현업에서도 그럴경우 crashlytics같이 분석할 수 있도록 툴에 쏴주던가 할때가 많아 굳이 이런 경우는 얼럿으로 가져가지 않아도 될것 같네요!
2. Custom Font에도 Dynamic Type을 적용
- 오.. 적절히 잘 구현해주셔서 접근성을 높였네요!
사실 접근성을 고려하지 않은 앱은 인종차별이나 마찬가지라고 어디서 유명한 포스팅을 봤었던 기억이..
근데 정말 저도 현업에서나 어디서나 그렇지만 이 접근성이 정말 애플의 강점이고 중요한데 프로젝트에 밀려 여유가 없다보니 고려를 잘 못하더라구요ㅠ
요게 사실 나라마다 다릅니다.
유럽이나 미국은 제가 알기론? 국가적으로 앱 검수 시 접근성이 없다면 리젝되는 곳도 있다고 알고 있어요.
(예를들어 웹페이지를 만들때 tap을 눌러 그 웹페이지 요소들의 이동 같은것이 없다면 리젝..)
반면 한국은 사실 접근성에 대해 국가적으로도 큰 고려의 대상이 아니기에 기업에서도 이걸 지켜서 얻는것보다 이 시간에 다른 프로젝트를 진행하는게 더 이율에 맞으니 계속 도태되고 있습니다ㅠ
국가적으로 접근성에 대해 구현이 없다면 리젝당하게 법으로 개정해야 될것 같은 부분이라고 봐요..
정말 중요한데 말처럼 쉽진 않네요ㅠ 이렇게 중요성을 알고 말하는 저도 못하고 있으니...😭
너무 사담이 길었네요!ㅋㅋ 결과적으로 잘 구현해주셔서 생각나서 넋두리 했습니다ㅎㅎ
3. 이미지 상세보기 시 보고있던 이미지가 그대로 넘어가도록 구현
- 보고 있던 이미지가 넘어가게 한 구현은 아주 좋습니다..!
그런데 리뷰 코멘트로도 달았지만 asyncAfter를 이용하여 강제적으로 delay를 주는것은 근본적인 해결은 아닙니다ㅠ
view 생명주기를 고려한다면 충분히 리팩토링 해보실 수 있을거에요ㅎㅎ
(리뷰 코멘트로 조금 더 자세히 남겼습니다!)
4. HIG를 참고하여 Action Sheet를 설계
- 너무너무 애플스럽습니다! 👍
5. CollectionView, PageControl을 이용한 이미지 Paging
- 이것도 너무너무 애플스럽고 기획보다 더 좋은 산출물이 나온것 같아요ㅋㅋ👍
6. 이미지 나타낼때 DispatchGroup을 사용
- 오호 이 구현도 좋네요ㅎㅎ
뷰 생명주기를 가지고도 해줄 수 있을거라 생각했는데 이 방법도 상당히 괜찮네요!
완료되기 전까지 시간이 길어지면 인디케이터를 넣어서 보여주면 더 퍼펙트하겠어요!
***
### 궁금한 점 및 조언이 필요한 부분에 대한 저의 의견
1. DetailViewController → ImageDetailViewController 로 전환할 때 viewDidLoad에서 collectionView.scrollToItem을 호출 시 제대로 적용되지 않아서 DispatchQueue.main.asyncAfter를 활용했는데, 적절한지 잘 모르겠습니다... 조언이 필요합니다!
- 이 부분에 대해 리뷰 코멘트로 자세히 남겨봤습니다..!
위에도 말한것 같지만 강제적 딜레이를 걸어주기 위해 임의적으로 asyncAfter를 사용하는건 좋지 않은것 같습니다🥲
뷰 생명주기를 다시 생각해보시면 해결될것 같습니다!
2. 네트워킹 하는 부분이 마치 산맥..처럼 공포의 들여쓰기가 생겨났는데, 이러한 부분을 어떻게 더 개선할 수 있을지 더이상 떠오르지가 않네요.. 저희가 할 수 있는 만큼 최대한 메소드를 분리해보았는데, 이대로도 괜찮을까요?
- 음.. 전 괜찮은데요..?
현업에서도 저 정도로 많이 들어가는 코드도 있고 최대한 분리하고 싶어도 어쩔수 없이 길어지는 코드가 있긴합니다.
저도 두분 입장을 가지고 봤을때 최대한 메서드로 분리해보신것 같아서 로직이나 기능에 문제가 되진 않는다 생각합니다.
사실 엄청난 들여쓰기 정도는 아닌것 같다고 생각했어요 전🥲
3. CollectionView에 FlowLayout을 따로 코드로 대입해주지 않으면, UICollectionViewDelegateFlowLayout를 채택하여 레이아웃을 설정해주어도 레이아웃이 적용되지 않는데, 왜그런걸까요?? 조언 부탁드립니다!
- UICollectionViewDelegateFlowLayout을 통해 설정이 적용될 FlowLayout 인스턴스가 없으니까 그렇지 않을까 생각해요🙋🏻
4. ScrollView를 활용해서 Zoom을 구현하였지만, 디바이스 전체 높이에 맞춰서 maximumZoomScale을 설정하는 방법을 모르겠습니다.
- 이해를 못했습니다!
아 혹시 이 말씀이신가요?
지금은 이미지의 2배수만큼 최대로 늘어날 수 있도록 해주셨는데 그러다보니 애초에 큰 이미지같은 경우 확대가 되었을때 화면에 다 담기지 않아서 최대로 확대해도 화면에 다 담겼으면 좋겠다..! 
그러기 위해 디바이스 높이만큼만 늘어났음 좋겠다 이 의미인가요?
그 의도가 맞으시다면 왜 그 의도를 택하시려는건지 의문이 듭니다🤔
보통의 사진 앨범(아이폰 앨범 등..)들에서 사진을 확대하면 지금 구현하신것과 같이 되어있는데 확대를 디바이스 사이즈만큼만 한정 지어주면
거의 확대가 되지 않는 느낌아닐까 해서요..!
5. 에러처리하는 부분을 일단 사용자에게 알릴 수 있는 얼럿을 제외하고는 네트워크 부분이나, 리퀘스트 부분은 print를 호출하도록 하였는데, 이 부분이 적절한걸까요? 부족하다면 조언 부탁드리겠습니다!
- 고민하셨던 1번하고 같은 맥락으로 보입니다ㅎㅎ
네 이건 얼럿을 꼭 안해도 됩니다! 지금같은 구현으로 사용자에게 불필요한 얼럿을 띄워주지 않는 이점을 가져가는 의도였다면 지금 구현이 더 좋습니다ㅎㅎ
고민 1번에서도 말씀드렸지만 후속처리만 잘 되어있다면 전~혀 문제없습니다👍
6. requestDelete 메소드 내부에서 success시 해당 데이터를 가지고 또 이어서 delete 메소드를 호출하여 삭제 요청을 진행하는 부분이 적절한 흐름인지 잘모르겠습니다.
- 논리적 결함은 없는것으로 보이는데 가독이 조금 떨어지긴 합니다ㅠ
삭제하려는 ID 조회 후 삭제 요청을 한 맥락으로 가져가는것 괜찮아 보입니다.
delete에서 어떠한 실패가 있었을때 저 isSuccess == false 로직이 타질텐데 사실 그에 대해서는 delete 내부에서
print(error.localizedDescription)로 처리했다고 보여서 굳이 필요할까라는 생각은 드네요.
말씀대로 아예 클라이언트의 요청 오류가 아니고 서버 내부적 500번대 오류거나 네트워크 단절 된 상태에서 호출 시 등에 해당 구문을 탈것 같아요.
그럴때를 대비해 print 로직을 추가해주는것도 괜찮아 보입니다!
***
다시 한번 너무 고생 많으셨습니다🙌
4주간 제 리뷰를 견뎌주시느라 답답한점도 막막한점도 있으셨을텐데 너무 두분의 실력이 나이스해서 잘 마무리가 되었던것 같아요ㅎㅎ
저도 덕분에 찐으로 한달동안 많이 배웠습니다 감사합니다🙇🏻
아 아직 이번주가 남았으니..!
필수 구현 외에도 추가 구현도 완벽한데 그래도 수정할 부분들이 있다고 판단은 되서 request change로 남겨두겠습니다!
수정되면 재요청 해주세요👍

llghdud921 committed:
feat: EditController내 사용자 입력값 검증을 실시간으로 도와주는 기능 추가 #22

- 유효하지 않은값을 나타내는 label 추가
- Editcontroller에 addTarget을 이용해서 실시간으로 사용자에게 알림을 주는 기능 구현

leeari95 committed:
refactor: 현재 보고있던 이미지를 가져오는 부분을 asyncAfter에서 viewWillAppear로 수정 #25

- 강제적 딜레이를 걸어주기 보다는 View Life Cycle을 활용하여 해당 문제 개선

llghdud921 committed:
refactor: zoom 기능 구현 코드내 불필요한 코드 제거 #25

leeari95 committed:
style: complection으로 오표기 되어있는 부분을 completion으로 모두 정정 #25

llghdud921 committed:
refactor: 변수, 메서드명 네이밍 개선 #25

leeari95 committed:
refactor: 누락된 접근제어 추가 및 메소드 순서 수정 #25

- setUpTextFields 메소드를 생성하여 관련 메소드를 한 곳에 모아두어 내부 개선

leeari95 commented:
안녕하세요! 그린~ 마지막까지도 꼼꼼하고 정성스런 피드백 정말 감사드립니다.
한달동안 같이 달려와주시느라 고생하셨습니다! 덕분에 많이 배우고 깨우치며 재밌는 한달을 보냈던 것 같아요.
이번에 피드백이 별로 없어서 저희가 정말 잘했나[?] 의문이 들기도 하고요.. 부족한게 많은거 같았는데,
잘했다고 해주시니 정말 감사할따름이네여... 😊
서론이 길었는데 이번에도 피드백 반영하여 수정사항은 아래와 같습니다~!

## 수정사항
* 현재 보고있던 이미지 가져오는 부분을 asyncAfter에서 viewWillAppear로 수정
* zoom 기능 구현 내부에 불필요한 코드 제거
* 오타 정정
* 네이밍 개선, 접근제어 추가, 메소드 순서 개선

## 추가 구현사항
* 사용자의 입력을 실시간으로 검증하여 Label로 사용자에게 검증 결과를 알려주는 기능을 추가 구현

이번에 서로 추가해보고 싶은 기능도 생겨서 한번 추가적으로 구현해보았습니다!
보시고 이상없으면 merge 부탁드리겠습니다.
감사했습니다!! 🙇🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/Controller/EditViewController.swift:
```diff
@@ -228,6 +244,69 @@ extension EditViewController {
+        } else {
+            DispatchQueue.main.async {
+                self.textFieldsStackView.nameTextField.layer.borderWidth = 0
+                self.textFieldsStackView.nameInvalidLabel.isHidden = true
```
아래 3개의 메서드에서 동일한 로직으로 보이는데 하나의 메서드로 분리시켜 인자 전달로 호출하면 어떨까요?🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/View/TextFieldsStackView.swift:
```diff
@@ -24,12 +24,17 @@ class TextFieldsStackView: UIStackView {
     @IBOutlet weak var stockTextField: UITextField!
     @IBOutlet weak var descriptionTextView: UITextView!
     
+    @IBOutlet weak var nameInvalidLabel: UILabel!
+    @IBOutlet weak var priceInvalidLabel: UILabel!
+    @IBOutlet weak var discountedPriceInvalidLabel: UILabel!
```
직역해볼때 조금 네이밍이 어색하다고 느껴졌습니다🥲
진짜 그냥 직역때리면 `이름/가격/할인가격유효하지않은레이블`인데 의도하셨던건 `유효하지않은 이름/가격/할인가격 레이블`이였을것 같아요.
그렇게 생각했을때도 해당 레이블이 조금 모호하지 않나 싶어요.
그렇게 생각한 이유는 우선 저 레이블들은 기존 이름/가격/할인가격 레이블을 가지고 유효하지 않으면 텍스트가 들어가서 알려주는 느낌이 들어요.
그래서 notify의 느낌이 들었습니다.
저도 딱히 생각나는 네이밍이 애매하지만... `announceForValidPriceLabel`혹은 `infoForValidPriceLabel`과 같은 느낌의
유효한 이름/가격/할인가격을 위한 안내 혹은 정보 레이블의 성향이 좀 더 맞지 않을까 하는 의견이 있습니다🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Views/Edit/View/TextFieldsStackView.swift:
```diff
@@ -85,6 +90,21 @@ class TextFieldsStackView: UIStackView {
         )
     }
     
+    func setUpInvaildLabel() {
```
만약 위 레이블의 네이밍이 변경된다면 이 메서드의 네이밍도 변경될 수 있을것 같아요..!
해당 메서드명만 봤을때는 어떤 레이블인지 이해하기가 조금 힘듭니다😭
유효하지 않은 레이블을 유추하기가 처음본다면 생각보다 쉽지 않을것 같았요.

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -144,11 +144,11 @@ class DetailViewController: UIViewController {
     }
     
-    private func deleteAction(_ secret: String) {
+    private func showDeleteSuccess(_ secret: String) {
```
두 메서드 로직 내에서  `AlertMessage.wrongPassword`의 잘못된 패스워드 얼럿 노출로 탈때도 있기에 변경하신 메서드 네이밍이 부적절해 보입니다🙋🏻
메서드 네이밍이 내부 로직을 포괄적으로 나타낼 수 있어야한다고 생각하는데 wrongPassword 얼럿이 노출되는건 해당 메서드 시그니쳐를 보고 전혀 유추할 수가 없습니다😭
기존에서 `modifyAction` 혹은 제안드렸던 `modifyProduct`를 생각해보면 수정된다고 포괄적인 네이밍으로 가져가고 그 안에서 성공하던 실패하던 각 분기를 태워서 적절하긴 합니다.
지금의 메서드명으로는 꼭 성공된 케이스만 보여지는걸로 해석이 됩니다!

GREENOVER commented on OpenMarket/OpenMarket/Views/Detail/Controller/DetailViewController.swift:
```diff
@@ -144,11 +144,11 @@ class DetailViewController: UIViewController {
             if isSuccess {
                 DispatchQueue.main.async {
-                    self.showAlert(message: AlertMessage.CompleteProductdelete) {
+                    self.showAlert(message: AlertMessage.completeProductdelete) {
```
`completeProductDelete`가 맞을것 같아요🙋🏻
delete -> Delete

GREENOVER reviewed (approved):
두분 너무 고생 하셨습니다!
정말 정말 두분이 잘해주셔서 이번 스텝에서는 리뷰가 별로 없었어요..!
어느새 코멘트가 많던거에 익숙해지셨나봅니다😅
수정된 사항 및 추가 구현 사항 확인했습니다!
요구사항에 만족하던건 한~참 전에 만족했고 추가 사항들까지 욕심내서 해주셨던 턱에 저도 더 욕심내서 보완되면 더 완벽할것 같은 부분에 리뷰를 드려봤씁니다🥲
확인 부탁드리겠습니다🙌
***
아 코드 리뷰와 별개로 당연히 이것도 요구사항은 아니지만.. 상용앱이라는 생각을 가지고 조금 사용자의 입장에서 봤을때 기능 이슈가 하나 있습니다🤔
현재 numberPad를 쓰는 상품가격/할인가격/수량 등에서 소수점을 여러개 찍어 상품 등록을 한다면 자세한 얼럿이 없고 상품정보 업데이트에 실패했습니다. 라고만 뜹니다.
아마 정상적이지 않은 데이터가 입혀지고 디코딩하다보니 실패했을것 같아요🤔
이에 2가지 방법이 있을것 같아요.
1. 계산기에서 하셨듯이..?! 소수점에 대해 검증하고 비정상적인 입력이 되지 않도록 textField 입력 시 방지
2. 상품등록 버튼 클릭 시 등록이 되지 않고 상세히 어떤 부분의 오류라는것을 안내
아마 1번의 방법은 코드 추가가 좀 더 필요하겠으며 2번은 지금 코드에서 간단히 넣긴 할 수 있을것 같습니다.
다만 근본적으로 올바른 길은 1번이긴 합니다!
관련해서 아래 해당 이슈 캡쳐 공유드립니다.
![Simulator Screen Shot - iPhone 11 - 2022-01-27 at 16 18 51](https://user-images.githubusercontent.com/72292617/151314151-3ab4c431-3e3a-4ced-92f4-8a3a69c817b8.png)
***
다시 한번 두분께 많이 배우고 재밌었던 4주였네요👍
감사합니다🙇🏻
어프로브 하겠습니다!
